### PR TITLE
Add static analysis

### PR DESCRIPTION
**Pull request by Mecharyry**
:x: ~Blocked until #283 is merged~ ❌ 

## Problem

We're not doing static analysis on this project...

## Solution

Add the standard static analysis setup, with rules. Includes the fixes from #283. Main fixes on this branch are in the `team-props` directory, which is the standard setup for static analysis.

### Paired with 

Nobody

### CODE REVIEW

**Comment by Mecharyry** [comment]
ℹ️  This includes the fixes from #282, mainly to look at anything in `team-props` but this contains the default setup for static analysis ℹ️

File: team-props/android-code-quality.gradle [hunk 1]
```diff
@@ -0,0 +1,13 @@
+task clean(type: Delete, dependsOn: subprojects.collect { p -> "$p.name:clean" }) {
```
Comment (Mecharyry) [inline team-props/android-code-quality.gradle]: Review this

File: team-props/static-analysis.gradle [hunk 2]
```diff
@@ -0,0 +1,37 @@
+apply plugin: 'com.novoda.static-analysis'
```
Comment (Mecharyry) [inline team-props/static-analysis.gradle]: Review this

File: team-props/static-analysis/checkstyle-modules.xml [hunk 3]
```diff
@@ -0,0 +1,163 @@
+<!DOCTYPE module PUBLIC
```
Comment (Mecharyry) [inline team-props/static-analysis/checkstyle-modules.xml]: Review this

File: team-props/static-analysis/checkstyle-suppressions.xml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0"?>
```
Comment (Mecharyry) [inline team-props/static-analysis/checkstyle-suppressions.xml]: Review this

File: team-props/static-analysis/findbugs-excludes.xml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<FindBugsFilter>
```
Comment (Mecharyry) [inline team-props/static-analysis/findbugs-excludes.xml]: Review this

File: team-props/static-analysis/lint-config.xml [hunk 6]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Mecharyry) [inline team-props/static-analysis/lint-config.xml]: Review this

File: team-props/static-analysis/pmd-rules.xml [hunk 7]
```diff
@@ -0,0 +1,142 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (Mecharyry) [inline team-props/static-analysis/pmd-rules.xml]: Review this

File: build.gradle [hunk 8]
```diff
@@ -1,3 +1,15 @@
+def teamPropsFile(propsFile) {
```
Comment (Mecharyry) [inline build.gradle]: Review this

File: demo-simple/build.gradle [hunk 9]
```diff
@@ -1,21 +1,22 @@
...
+        targetSdkVersion 27
         versionCode 1
         versionName "1.0"
         testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
     }
 
     lintOptions {
-        abortOnError false
-        lintConfig file("$rootProject.projectDir/team-props/lint-rules/config.xml")
+        lintConfig teamPropsFile('static-analysis/lint-config.xml')
```
Comment (danybony) [inline demo-simple/build.gradle]: 💯

File: library/build.gradle [hunk 10]
```diff
@@ -2,21 +2,22 @@ apply plugin: 'com.android.library'
...
 
         testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
     }
 
     lintOptions {
-        abortOnError false
-        lintConfig file("$rootProject.projectDir/team-props/lint-rules/config.xml")
+        lintConfig teamPropsFile('static-analysis/lint-config.xml')
+        abortOnError true
+        warningsAsErrors true
```
Comment (danybony) [inline library/build.gradle]: 👍

File: team-props/static-analysis/pmd-rules.xml [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    <exclude name="ShortClassName" />
+    <exclude name="AvoidFieldNameMatchingMethodName" />
+  </rule>
+
+  <rule ref="rulesets/java/codesize.xml">
+    <exclude name="TooManyMethods" />
+    <exclude name="TooManyFields" />
+  </rule>
+
+  <!-- Suppress ExcessiveParameterList check for Api models ctors and factory methods -->
```
Comment (joetimmins) [inline team-props/static-analysis/pmd-rules.xml]: Do we actually have any classes that start with `Api` in this project?

File: team-props/static-analysis/pmd-rules.xml [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+  <rule ref="rulesets/java/optimizations.xml">
+    <exclude name="LocalVariableCouldBeFinal" />
+    <exclude name="MethodArgumentCouldBeFinal" />
+    <exclude name="AvoidInstantiatingObjectsInLoops" />
+  </rule>
+  <rule ref="rulesets/java/basic.xml" />
+  <rule ref="rulesets/java/sunsecure.xml" />
+  <rule ref="rulesets/java/coupling.xml">
+    <exclude name="LoosePackageCoupling" />
+    <!-- LawOfDemeter seems to be bugged -->
```
Comment (joetimmins) [inline team-props/static-analysis/pmd-rules.xml]: sadface. i like law of demeter as an idea

File: team-props/android-code-quality.gradle (hunk 1 above)
Comment (joetimmins) [reply team-props/android-code-quality.gradle]: Seems legit

File: team-props/static-analysis/checkstyle-modules.xml [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+    <module name="SimplifyBooleanExpression" />
+    <module name="SimplifyBooleanReturn" />
+
+    <!-- Checks for class design                         -->
+    <!-- See http://checkstyle.sf.net/config_design.html -->
+    <module name="FinalClass" />
+    <module name="HideUtilityClassConstructor" />
+    <module name="InterfaceIsType" />
+    <module name="VisibilityModifier">
+      <!-- Ignore fields injected by Dagger as it requires at least package private visibility -->
```
Comment (joetimmins) [inline team-props/static-analysis/checkstyle-modules.xml]: Do we have Dagger on this project?

File: team-props/static-analysis/pmd-rules.xml (hunk 11 above)
Comment (Mecharyry) [reply team-props/static-analysis/pmd-rules.xml]: 🤔 I don't think so, we can probably kill this one.

File: team-props/static-analysis/lint-config.xml [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+<lint>
+
+  <issue id="MissingTranslation" severity="ignore" />
+  <issue id="GoogleAppIndexingWarning" severity="ignore" />
+
+  <issue id="OldTargetApi" severity="informational" />
+
+  <!-- Ignore RtL warnings since the app doesn't support RtL -->
```
Comment (joetimmins) [inline team-props/static-analysis/lint-config.xml]: DownloadManager might be used in an app that does support RtL no?

Comment (Mecharyry) [reply team-props/static-analysis/lint-config.xml]: This only applies to this codebase, client applications would need to reference this `lint-config` explicitly for it to be picked up by their buildscript. At least, that's how I understand it 😄

File: team-props/static-analysis/checkstyle-modules.xml (hunk 13 above)
Comment (Mecharyry) [reply team-props/static-analysis/checkstyle-modules.xml]: Ah nope 😄  will remove.

File: team-props/static-analysis/pmd-rules.xml (hunk 11 above)
Comment (Mecharyry) [reply team-props/static-analysis/pmd-rules.xml]: 690c984

File: team-props/static-analysis/checkstyle-modules.xml (hunk 13 above)
Comment (Mecharyry) [reply team-props/static-analysis/checkstyle-modules.xml]: 7c086ec

**Pull request merged**